### PR TITLE
second round of pybind11 changes

### PR DESCRIPTION
**Pull request by tgn3000**
- Do not use `PyArray_SimpleNew` for pybind11, as it causes memory leaks. Hence also avoid using pybind11::to_array_astype<T>(PyObject*).
- Add specialized template function for `pybind11::to_object(const T& s)` with T = std::string, and use `pybind11::bytes(s)` instead of `pybind11::cast(s)`. This fixes utf8 decoding errors when storing trajectories in binary form.
- Add `openravepy::toPyArray` for `std::array`.
- Use v.data() [resp. v] instead of &v[0], if v is a STL container [resp. C array].
- Use v.empty() instead of comparing v.size() with 0.
- Add LinkInfo::_bVisible, which only occurs once in planningcommon. Not sure if it's intended.
- Move class function implementations from header to cpp files.
- Other small fixes/improvements.

### CODE REVIEW

**Comment by tgn3000** [comment]
@rdiankov  ping. Merged today's production after your `jsonserialization2_infos` merge.

File: python/bindings/include/openravepy/openravepy_jointinfo.h [hunk 1]
```diff
@@ -36,12 +36,20 @@ class PySideWall
     int type = 0;
 };
 
+class PyTestPickle
```
Comment (rdiankov) [inline python/bindings/include/openravepy/openravepy_jointinfo.h]: @tgn3000 is this class necessary?

File: CMakeLists.txt [hunk 2]
```diff
@@ -862,7 +861,7 @@ if(OPT_PYTHON AND PYTHON_EXECUTABLE)
   set(OPENRAVEPY_VER_INSTALL_DIR "${OPENRAVEPY_INSTALL_DIR}/${OPENRAVEPY_VER_NAME}") # used by openrave-config.cmake
 
   # build python bindings, either with pybind11 or Boost.Python
-  set(USE_PYBIND11_PYTHON_BINDINGS OFF CACHE BOOL "Build OpenRAVE's python bindings with pybind11 (ON) or Boost.Python (OFF)")
+  set(USE_PYBIND11_PYTHON_BINDINGS ON CACHE BOOL "Build OpenRAVE's python bindings with pybind11 (ON) or Boost.Python (OFF)")
```
Comment (rdiankov) [inline CMakeLists.txt]: @tgn3000 for now, can you make the default OFF?

File: python/bindings/include/openravepy/openravepy_int.h [hunk 3]
```diff
@@ -331,18 +331,42 @@ inline RaveTransformMatrix<T> ExtractTransformMatrixType(const py::object& o)
...
     return py::to_array_astype<dReal>(pyvalues);
+#endif // USE_PYBIND11_PYTHON_BINDINGS
 }
 
 template <typename T>
 inline py::object toPyArray3(const std::vector<RaveVector<T> >& v)
 {
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    const size_t numel = v.size() * 3;
+    std::vector<T> arr(numel);
```
Comment (rdiankov) [inline python/bindings/include/openravepy/openravepy_int.h]: @tgn3000 is there a way to do this without creating a new `vector`?

File: python/bindings/openravepy_collisionchecker.cpp [hunk 4]
```diff
@@ -524,11 +524,20 @@ object PyCollisionCheckerBase::CheckCollisionRays(object rays, PyKinBodyPtr pbod
     CollisionReportPtr preport(&report,null_deleter());
 
     RAY r;
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    const size_t numel = num * 6;
+    std::vector<dReal> vpos(numel);
```
Comment (rdiankov) [inline python/bindings/openravepy_collisionchecker.cpp]: @tgn3000 is there a way to do this without creating a new `vector`?

File: python/bindings/openravepy_global.cpp [hunk 5]
```diff
@@ -248,6 +248,27 @@ class PyTriMesh
     PyTriMesh(object vertices, object indices) : vertices(vertices), indices(indices) {
     }
     PyTriMesh(const TriMesh& mesh) {
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+        // vertices
+        const int nvertices = mesh.vertices.size();
+        const int numel = nvertices * 3;
+        std::vector<dReal> vvertices(numel);
```
Comment (rdiankov) [inline python/bindings/openravepy_global.cpp]: @tgn3000 is there a way to do this without creating a new `vector`?

File: python/bindings/openravepy_global.cpp [hunk 6]
```diff
@@ -1025,13 +1047,18 @@ object poseFromMatrix(object o)
 
 object poseFromMatrices(object otransforms)
 {
-    int N = len(otransforms);
+    const int N = len(otransforms);
     if( N == 0 ) {
         return py::empty_array_astype<dReal>();
     }
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    std::vector<dReal> vvalues(N * 7);
```
Comment (rdiankov) [inline python/bindings/openravepy_global.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_global.cpp [hunk 7]
```diff
@@ -1041,31 +1068,49 @@ object poseFromMatrices(object otransforms)
...
 
 object InvertPoses(object o)
 {
-    int N = len(o);
+    const int N = len(o);
     if( N == 0 ) {
         return py::empty_array_astype<dReal>();
     }
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    std::vector<dReal> vtrans(N * 7);
```
Comment (rdiankov) [inline python/bindings/openravepy_global.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_global.cpp [hunk 8]
```diff
@@ -1114,16 +1159,27 @@ object poseTransformPoint(object opose, object opoint)
 
 object poseTransformPoints(object opose, object opoints)
 {
-    Transform t = ExtractTransformType<dReal>(opose);
-    int N = len(opoints);
+    const Transform t = ExtractTransformType<dReal>(opose);
+    const int N = len(opoints);
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    std::vector<dReal> vtrans(N * 3);
```
Comment (rdiankov) [inline python/bindings/openravepy_global.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_int.cpp [hunk 9]
```diff
@@ -1306,19 +1326,28 @@ object PyEnvironmentBase::CheckCollisionRays(py::numeric::array rays, PyKinBodyP
...
 
-    bool isFloat = PyArray_ITEMSIZE(pPyRays) == sizeof(float); // or double
+    const bool isFloat = PyArray_ITEMSIZE(pPyRays) == sizeof(float); // or double
     const float *pRaysFloat = isFloat ? reinterpret_cast<const float*>(PyArray_DATA(pPyRays)) : NULL;
     const double *pRaysDouble = isFloat ? NULL : reinterpret_cast<const double*>(PyArray_DATA(pPyRays));
 
     RAY r;
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    const size_t numel = nRays * 6;
+    std::vector<dReal> vpos(numel);
```
Comment (rdiankov) [inline python/bindings/openravepy_int.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_kinbody.cpp [hunk 10]
```diff
@@ -87,26 +87,34 @@ void PySideWall::Get(KinBody::GeometryInfo::SideWall& sidewall) {
...
-    _bVisible = true;
-    _bModifiable = true;
+// TGN: debug; will remove PyTestPickle
+PyTestPickle::PyTestPickle() {
+    if(IS_PYTHONOBJECT_NONE(_arr)) {
+        const std::vector<dReal> v {1, 2, 4.5, -3.0, 8, 9, 10, 11, 12};
+        _arr = toPyArray(v);
+    }
 }
+PyTestPickle::~PyTestPickle() {}
```
Comment (rdiankov) [inline python/bindings/openravepy_kinbody.cpp]: @tgn3000 why is this class necessary?

File: python/bindings/openravepy_kinbody.cpp [hunk 11]
```diff
@@ -2620,24 +2721,36 @@ void PyKinBody::SetDOFVelocities(object odofvelocities, uint32_t checklimits, ob
...
+    if( _pbody->GetLinks().empty() ) {
         return py::empty_array_astype<dReal>();
     }
     std::vector<std::pair<Vector,Vector> > velocities;
     _pbody->GetLinkVelocities(velocities);
-
+    const size_t nvelocities = velocities.size();
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    const size_t numel = nvelocities * 6;
+    std::vector<dReal> vvelocities(numel);
```
Comment (rdiankov) [inline python/bindings/openravepy_kinbody.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_kinbody.cpp [hunk 12]
```diff
@@ -2672,9 +2785,16 @@ object PyKinBody::GetLinkAccelerations(object odofaccelerations, object oexterna
     std::vector<std::pair<Vector,Vector> > vLinkAccelerations;
     _pbody->GetLinkAccelerations(vDOFAccelerations, vLinkAccelerations, pmapExternalAccelerations);
 
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    const size_t nLinkAccelerations = vLinkAccelerations.size();
+    const size_t numel = nLinkAccelerations * 6;
+    std::vector<dReal> vaccel(numel);
```
Comment (rdiankov) [inline python/bindings/openravepy_kinbody.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_robot.cpp [hunk 13]
```diff
@@ -559,15 +559,29 @@ object PyRobotBase::PyManipulator::FindIKSolutions(object oparam, int filteropti
             return py::empty_array_astype<dReal>();
         }
 
-        npy_intp dims[] = { npy_intp(vsolutions.size()), npy_intp(_pmanip->GetArmIndices().size()) };
+        const size_t nSolutions = vsolutions.size();
+        const size_t nArmIndices = _pmanip->GetArmIndices().size();
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+        const size_t numel = nSolutions * nArmIndices;
+        std::vector<dReal> vpos(numel);
```
Comment (rdiankov) [inline python/bindings/openravepy_robot.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_robot.cpp [hunk 14]
```diff
@@ -606,15 +620,29 @@ object PyRobotBase::PyManipulator::FindIKSolutions(object oparam, object freepar
...
```
Comment (rdiankov) [inline python/bindings/openravepy_robot.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/openravepy_trajectory.cpp [hunk 15]
```diff
@@ -255,17 +275,37 @@ object PyTrajectoryBase::__getitem__(py::slice indices) const
...
     _ptrajectory->GetWaypoint(0,values);
-    int numdof = _ptrajectory->GetConfigurationSpecification().GetDOF();
-    npy_intp dims[] = { npy_intp(vindices.size()), npy_intp(numdof) };
+    const int nindices = vindices.size();
+    const int numdof = _ptrajectory->GetConfigurationSpecification().GetDOF();
+    const int nvalues = values.size();
+    OPENRAVE_ASSERT_OP(numdof, ==, nvalues);
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    const int numel = nindices * numdof;
+    std::vector<dReal> vpos(numel);
```
Comment (rdiankov) [inline python/bindings/openravepy_trajectory.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/pyann.cpp [hunk 16]
```diff
@@ -159,27 +166,39 @@ object search(ANNkd_tree& kdtree, object q, int k, double eps, bool priority = f
...
     if( N == 0 ) {
         return py::make_tuple(py::empty_array_astype<int>(), py::empty_array_astype<ANNdist>());
     }
 
     BOOST_ASSERT(len(qarray[0])==kdtree.theDim());
     ANNpointManaged annq(kdtree.theDim());
+
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    std::vector<ANNdist> vdists(N * k);
+    std::vector<int> vidx(N * k);
```
Comment (rdiankov) [inline python/bindings/pyann.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/pyann.cpp [hunk 17]
```diff
@@ -190,45 +209,65 @@ object search_array(ANNkd_tree& kdtree, object qarray, int k, double eps, bool p
...
     if( kball <= 0 ) {
         return py::make_tuple(py::empty_array_astype<int>(),py::empty_array_astype<ANNdist>(),kball);
     }
-
-    npy_intp dims[] = {std::min(k, kball)};
+    
+    const int numel = std::min(k, kball);
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    std::vector<ANNdist> vdists(numel);
+    std::vector<int> vidx(numel);
```
Comment (rdiankov) [inline python/bindings/pyann.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/pyann.cpp [hunk 18]
```diff
@@ -248,33 +288,55 @@ object k_fixed_radius_search(ANNkd_tree& kdtree, object q, double sqRad, int k,
...
-    if( N == 0 )
+    const int N = len(qarray);
+    if( N == 0 ) {
         return py::make_tuple(py::empty_array_astype<int>(), py::empty_array_astype<ANNdist>(), py::empty_array_astype<int>());
+    }
 
     BOOST_ASSERT(len(qarray[0])==kdtree.theDim());
     ANNpointManaged annq(kdtree.theDim());
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    std::vector<int> vkball(N);
```
Comment (rdiankov) [inline python/bindings/pyann.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/pyann.cpp [hunk 19]
```diff
@@ -248,33 +288,55 @@ object k_fixed_radius_search(ANNkd_tree& kdtree, object q, double sqRad, int k,
...
+        return py::make_tuple(py::empty_array_astype<int>(), py::empty_array_astype<ANNdist>(), openravepy::toPyArray(vkball));
+#else
         return py::make_tuple(py::empty_array_astype<int>(), py::empty_array_astype<ANNdist>(), py::to_array_astype<int>(pykball));
+#endif // USE_PYBIND11_PYTHON_BINDINGS
     }
 
+#ifdef USE_PYBIND11_PYTHON_BINDINGS
+    const size_t numel = N * k;
+    std::vector<ANNdist> vdists(numel);
+    std::vector<int> vidx(numel);
```
Comment (rdiankov) [inline python/bindings/pyann.cpp]: @tgn3000 is there a way to do this without creating a new vector?

File: python/bindings/include/openravepy/openravepy_jointinfo.h (hunk 1 above)
Comment (tgn3000) [reply python/bindings/include/openravepy/openravepy_jointinfo.h]: no, these were merged from `pybind11_stable` branch by mistake. Removed.

File: CMakeLists.txt (hunk 2 above)
Comment (tgn3000) [reply CMakeLists.txt]: Merged from `pybind11_stable` branch by mistake. Set OFF.

File: python/bindings/openravepy_kinbody.cpp (hunk 10 above)
Comment (tgn3000) [reply python/bindings/openravepy_kinbody.cpp]: removed

File: python/bindings/include/openravepy/openravepy_int.h (hunk 3 above)
Comment (tgn3000) [reply python/bindings/include/openravepy/openravepy_int.h]: yes, done; same for the rest

File: python/bindings/openravepy_global.cpp (hunk 5 above)
Comment (tgn3000) [reply python/bindings/openravepy_global.cpp]: yes, done; same for the rest

**Comment by rdiankov** [comment]
looks good, thanks!

**Pull request merged**